File: frontend/src/lib/api/templates.ts
/**
 * Typed client for the templates REST surface (G7;
 * ``src/bearings/web/routes/templates.py``).
 *
 * Mirrors :class:`bearings.web.models.templates.TemplateOut` /
 * :class:`bearings.web.models.templates.TemplateIn` /
 * :class:`bearings.web.models.templates.TemplatePatch` /
 * :class:`bearings.web.models.templates.TemplateInstantiateIn` field for field.
 * The new-session dialog fetches the list on mount to populate the template
 * picker dropdown; the session-row context menu uses ``createTemplate`` to
 * implement the ``save_as_template`` action; ``instantiateTemplate`` creates a
 * fully-populated session from a template in one server-side transaction
 * (gap-cycle-13-006).
 */
import { API_TEMPLATES_ENDPOINT, templateEndpoint, templateInstantiateEndpoint } from "../config";
import { deleteResource, getJson, patchJson, postJson, type RequestOptions } from "./client";
import type { SessionOut } from "./sessions";

/** Wire shape — one-to-one with :class:`bearings.web.models.templates.TemplateOut`. */
export interface TemplateOut {
  id: number;
  name: string;
  description: string | null;
  model: string;
  advisor_model: string | null;
  advisor_max_uses: number;
  effort_level: string;
  permission_profile: string;
  system_prompt_baseline: string | null;
  working_dir_default: string | null;
  tag_names: string[];
  created_at: string;
  updated_at: string;
}

interface CreateTemplateParams {
  name: string;
  model: string;
  description?: string | null;
  advisor_model?: string | null;
  advisor_max_uses?: number;
  effort_level?: string;
  permission_profile?: string;
  system_prompt_baseline?: string | null;
  working_dir_default?: string | null;
  tag_names?: string[];
}

interface PatchTemplateParams {
  name?: string;
  model?: string;
  description?: string | null;
  advisor_model?: string | null;
  advisor_max_uses?: number;
  effort_level?: string;
  permission_profile?: string;
  system_prompt_baseline?: string | null;
  working_dir_default?: string | null;
  tag_names?: string[];
}

/**
 * Create a new template. Returns the created row.
 *
 * 409 when the name is already taken; 422 when routing fields are invalid.
 */
export async function createTemplate(
  params: CreateTemplateParams,
  options: RequestOptions = {},
): Promise<TemplateOut> {
  return await postJson<TemplateOut>(API_TEMPLATES_ENDPOINT, params, options);
}

/**
 * List all templates, alphabetically by name. Returns ``[]`` when none
 * exist yet — the picker renders only the "-- no template --" placeholder.
 */
export async function listTemplates(options: RequestOptions = {}): Promise<TemplateOut[]> {
  return await getJson<TemplateOut[]>(API_TEMPLATES_ENDPOINT, options);
}

/**
 * Fetch a single template by id. 404 surfaces as :class:`ApiError`.
 *
 * Not yet consumed by any UI component — export added when a template
 * detail / edit surface lands (future item).
 */
async function getTemplate(templateId: number, options: RequestOptions = {}): Promise<TemplateOut> {
  return await getJson<TemplateOut>(templateEndpoint(templateId), options);
}

/**
 * Partially update a template. Only the supplied fields are changed;
 * all others are preserved. 404 / 409 surface as :class:`ApiError`.
 *
 * Not yet consumed by any UI component — export added when a template
 * edit surface lands (future item).
 */
async function patchTemplate(
  templateId: number,
  params: PatchTemplateParams,
  options: RequestOptions = {},
): Promise<TemplateOut> {
  return await patchJson<TemplateOut>(templateEndpoint(templateId), params, options);
}

/**
 * Delete a template by id. 204 on success; 404 surfaces as :class:`ApiError`.
 */
export async function deleteTemplate(
  templateId: number,
  options: RequestOptions = {},
): Promise<void> {
  return await deleteResource<void>(templateEndpoint(templateId), options);
}

/** Optional override fields for :func:`instantiateTemplate`. */
interface InstantiateTemplateParams {
  /** Overrides the title; defaults to the template name server-side. */
  title?: string;
  /** Overrides the executor model. */
  model?: string;
  /** Overrides the description. */
  description?: string | null;
  /** Overrides the working directory. */
  working_dir?: string | null;
  /** Overrides the session instructions (baseline: system_prompt_baseline). */
  session_instructions?: string | null;
  /** Overrides the permission mode. */
  permission_mode?: string | null;
  /** Overrides the advisor model. */
  advisor_model?: string | null;
  /** Overrides the advisor max uses. */
  advisor_max_uses?: number;
  /** Overrides the effort level. */
  effort_level?: string;
}

/**
 * Create a new session from a template (gap-cycle-13-006).
 *
 * Calls ``POST /api/templates/{id}/instantiate`` which copies all template
 * fields (model, advisor settings, effort_level, permission_profile,
 * working_dir_default, system_prompt_baseline → session_instructions,
 * tag_names → attached tags, description) into a new session row in one
 * server-side sequence. The optional ``params`` body can override any field.
 *
 * 404 when the template does not exist; 422 when no working_dir is
 * resolvable or routing fields are invalid.
 */
export async function instantiateTemplate(
  templateId: number,
  params: InstantiateTemplateParams = {},
  options: RequestOptions = {},
): Promise<SessionOut> {
  return await postJson<SessionOut>(templateInstantiateEndpoint(templateId), params, options);
}

// Silence unused-variable warnings in editors until consumers are added.
void getTemplate;
void patchTemplate;

File: src/bearings/web/models/templates.py
# mypy: disable-error-code=explicit-any
"""Pydantic wire shapes for ``routes/templates.py`` (G7).

Per ``docs/architecture-v1.md`` §1.1.5 the wire DTOs live alongside
their route module. The shapes mirror :class:`bearings.db.templates.Template`
plus a :class:`TemplateIn` create body, :class:`TemplatePatch` update body,
and :class:`TemplateInstantiateIn` instantiate override body.

The ``mypy: disable-error-code=explicit-any`` pragma is the same narrow
carve-out the other ``web/models/*.py`` files make for Pydantic's
metaclass-exposed ``Any`` surface — every public ``BaseModel`` subclass
below has a fully-typed field set.
"""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field

from bearings.config.constants import (
    DEFAULT_TEMPLATE_ADVISOR_MAX_USES,
    DEFAULT_TEMPLATE_EFFORT_LEVEL,
    DEFAULT_TEMPLATE_PERMISSION_PROFILE,
    SESSION_DESCRIPTION_MAX_LENGTH,
    SESSION_TITLE_MAX_LENGTH,
    TEMPLATE_DESCRIPTION_MAX_LENGTH,
    TEMPLATE_NAME_MAX_LENGTH,
)


class TemplateIn(BaseModel):
    """Request body for ``POST /api/templates``.

    The DB layer's :func:`bearings.db.templates.create` validates the
    routing-alphabet fields in :class:`bearings.db.templates.Template.__post_init__`;
    a bad payload surfaces as a 422 from the route.
    ``model_config`` forbids extra fields so a typo (e.g. ``permissionProfile``
    instead of ``permission_profile``) is rejected at the wire boundary.
    """

    model_config = ConfigDict(extra="forbid")

    name: str = Field(min_length=1, max_length=TEMPLATE_NAME_MAX_LENGTH)
    model: str = Field(min_length=1)
    description: str | None = Field(
        default=None,
        max_length=TEMPLATE_DESCRIPTION_MAX_LENGTH,
    )
    advisor_model: str | None = None
    advisor_max_uses: int = Field(default=DEFAULT_TEMPLATE_ADVISOR_MAX_USES, ge=0)
    effort_level: str = DEFAULT_TEMPLATE_EFFORT_LEVEL
    permission_profile: str = DEFAULT_TEMPLATE_PERMISSION_PROFILE
    system_prompt_baseline: str | None = None
    working_dir_default: str | None = None
    tag_names: list[str] = Field(default_factory=list)


class TemplatePatch(BaseModel):
    """Request body for ``PATCH /api/templates/{id}``.

    Every field is optional; missing fields are filled from the existing
    row so the caller only has to send the delta. The route layer fetches
    the existing row, merges, and calls :func:`bearings.db.templates.update`
    with the full field set.
    """

    model_config = ConfigDict(extra="forbid")

    name: str | None = Field(default=None, min_length=1, max_length=TEMPLATE_NAME_MAX_LENGTH)
    model: str | None = Field(default=None, min_length=1)
    description: str | None = None
    advisor_model: str | None = None
    advisor_max_uses: int | None = Field(default=None, ge=0)
    effort_level: str | None = None
    permission_profile: str | None = None
    system_prompt_baseline: str | None = None
    working_dir_default: str | None = None
    tag_names: list[str] | None = None


class TemplateOut(BaseModel):
    """Response body for template endpoints.

    One-to-one with :class:`bearings.db.templates.Template`.
    """

    model_config = ConfigDict(extra="forbid")

    id: int
    name: str
    description: str | None
    model: str
    advisor_model: str | None
    advisor_max_uses: int
    effort_level: str
    permission_profile: str
    system_prompt_baseline: str | None
    working_dir_default: str | None
    tag_names: list[str]
    created_at: str
    updated_at: str


class TemplateInstantiateIn(BaseModel):
    """Request body for ``POST /api/templates/{id}/instantiate`` (gap-cycle-13-006).

    All fields are optional caller-side overrides. Omitted fields inherit
    the template's stored values:

    * ``title`` — defaults to the template name.
    * ``model`` — defaults to ``template.model``.
    * ``description`` — defaults to ``template.description``.
    * ``working_dir`` — defaults to ``template.working_dir_default``.
    * ``session_instructions`` — defaults to ``template.system_prompt_baseline``.
    * ``permission_mode`` — defaults to ``template.permission_profile``
      (empty string in the template maps to ``None``).
    * ``advisor_model`` / ``advisor_max_uses`` / ``effort_level`` —
      default to the corresponding ``template.advisor_*`` / ``template.effort_level`` fields.

    The route validates routing-alphabet fields via
    :class:`Session.__post_init__`; a bad override surfaces as 422.
    ``model_config`` forbids extra fields so a typo is caught at the wire
    boundary.
    """

    model_config = ConfigDict(extra="forbid")

    title: str | None = Field(default=None, min_length=1, max_length=SESSION_TITLE_MAX_LENGTH)
    model: str | None = Field(default=None, min_length=1)
    description: str | None = Field(default=None, max_length=SESSION_DESCRIPTION_MAX_LENGTH)
    working_dir: str | None = Field(default=None, min_length=1)
    session_instructions: str | None = None
    permission_mode: str | None = None
    advisor_model: str | None = None
    advisor_max_uses: int | None = Field(default=None, ge=0)
    effort_level: str | None = None


__all__ = ["TemplateIn", "TemplateInstantiateIn", "TemplateOut", "TemplatePatch"]
